### PR TITLE
[test] Add test to make sure ndarrays are initialized as zero

### PR DESCRIPTION
**Pull request by ailzhang**
Issue: #6438 

### Brief Summary
This PR only enables zero fill for ndarrays created from Python frontend so it doesn't affect performance of ndarray created from C-API. 
In the future we may provide an option for python users to skip the zero fill as needed but that will be a separate PR.

### CODE REVIEW

File: python/taichi/lang/_ndarray.py [hunk 1]
```diff
@@ -225,7 +225,7 @@ def __init__(self, dtype, arr_shape):
         super().__init__()
         self.dtype = cook_dtype(dtype)
         self.arr = impl.get_runtime().prog.create_ndarray(
-            self.dtype, arr_shape)
+            self.dtype, arr_shape, Layout.NULL, True)
```
Comment (lin-hitonami) [inline python/taichi/lang/_ndarray.py:228]:
```suggestion
            self.dtype, arr_shape, Layout.NULL, zero_fill=True)
```

File: taichi/program/program.cpp [hunk 2]
```diff
@@ -455,8 +455,25 @@ std::size_t Program::get_snode_num_dynamically_allocated(SNode *snode) {
...
 Ndarray *Program::create_ndarray(const DataType type,
                                  const std::vector<int> &shape,
-                                 ExternalArrayLayout layout) {
-  ndarrays_.emplace_back(std::make_unique<Ndarray>(this, type, shape, layout));
+                                 ExternalArrayLayout layout,
+                                 bool zero_fill) {
+  auto arr = std::make_unique<Ndarray>(this, type, shape, layout);
+  if (zero_fill) {
+    Arch arch = this_thread_config().arch;
+    if (arch == Arch::x64 || arch == Arch::cuda) {
```
Comment (lin-hitonami) [inline taichi/program/program.cpp:463]: or arch == Arch::arm64? I guess it will fail on M1 because arm64 uses CpuDevice as well. Maybe we can use `arch_is_cpu(arch)` here.

Comment (ailzhang) [reply taichi/program/program.cpp:463]: good point!

**Review by lin-hitonami — APPROVED** [review]
LGTM
